Refactor Move by Taxonomy box

mariadanieldeepak opened the pull request:
Fix #43

Close #44, close #65 (since those changes are now part of this PR, and are redundant in isolated fashion)   

**Note:**
The refactor part is completed and I'm writing the test cases.

mariadanieldeepak committed:
Refactor Move by Taxonomy box

sudar commented:
@mariadanieldeepak 

This branch has some conflicts. Can you resolve them so that I can do the review?

mariadanieldeepak committed:
Merge branch 'dev/2.0.0' into 43-feature/refactor-move-by-taxonomy

mariadanieldeepak commented:
@sudar,

> This branch has some conflicts. Can you resolve them so that I can do the review?

Conflicts have been resolved.

sudar commented on include/Page/Post.php:
```diff
@@ -14,6 +14,7 @@ protected function initialize() {
 		$this->page_title      = __( 'Bulk Move Posts', 'bulk-move' );
 		$this->menu_title      = __( 'Bulk Move Posts', 'bulk-move' );
 		$this->warning_message = __( 'WARNING: Posts moved once cannot be reverted. Use with caution.', 'bulk-move' );
+
```
I am seeing lot of blank lines getting added everywhere. Let's avoid it. It is going to generate noise in diffs

mariadanieldeepak replied:
> I am seeing lot of blank lines getting added everywhere. Let's avoid it. It is going to generate noise in diffs

Fixed.

sudar commented on bulk-move.php:
```diff
@@ -66,6 +66,7 @@ function bm_autoloader( $class_name ) {
 function load_bulk_move() {
 	$bulk_move = bulk_move();
 	$bulk_move->set_plugin_file( __FILE__ );
+	$bulk_move->add_loadie( new BM_Request_CustomTaxonomyAction() );
```
The Ajax handlers should be the responsibility of the Metabox, not the main plugin class.

sudar replied:
Once the Ajax handlers are moved to the Metabox module, the Loadie interface will not be needed. Let's move that code to a new branch and keep it.

If we ever need a Loadie interface then we can merge it into master.

mariadanieldeepak replied:
@Sudar,

> The Ajax handlers should be the responsibility of the Metabox, not the main plugin class.

This is so true. I initially added the AJAX requests as part of the metabox, and found that the response from the AJAX request was always 0. I found that this was because that the callback couldn't be found by the `admin-ajax.php` file.

I'll move back the code and make a commit, for you to refer.

sudar commented on include/Page/Base.php:
```diff
@@ -298,7 +299,7 @@ public function add_metabox( $metabox ) {
 		}
 
-		$this->metaboxes[] = $metabox;
+		$this->metaboxes[ $metabox->get_metabox_slug() ] = $metabox;
```
Any specific reason for adding the key to the array?

mariadanieldeepak replied:
> Any specific reason for adding the key to the array?

I actually tried to hook the AJAX requests in the Metabox class initially. At that point, the AJAX requests didn't work and move them to a separate file. But then, I forgot to remove these methods. Now I've removed them.

sudar commented on include/Page/Base.php:
```diff
@@ -236,7 +237,7 @@ public function enqueue_script() {
 		$bulk_move_posts = array(
```
All these are responsibility of the Custom Taxonomy metabox or the Ajax handlers.

Create a filter and then use the filter to add these values from inside those classes.

sudar commented on include/Metabox/Posts/CustomTaxonomy.php:
```diff
@@ -0,0 +1,161 @@
+
+		</fieldset>
+
+		<p class="submit bm_ct_submit">
```
Use the `render_submit` helper for this.

mariadanieldeepak replied:
> Use the `render_submit` helper for this.

Fixed

sudar commented on include/Metabox/Posts/CustomTaxonomy.php:
```diff
@@ -0,0 +1,161 @@
+				</tr>
+
+				<tr class="term-select-row">
+					<td scope="row" >
```
Abstract this into a helper function that is added to the base class

mariadanieldeepak replied:
@sudar,
 
> Abstract this into a helper function that is added to the base class

The Select Term dropdown has only the `<select>` and a default `<option>` tag. Even if we abstract it out to the the base class, it will be dependent on the AJAX to load terms.

Do you think we should restructure the AJAX JS, or should we keep this Terms `<select>` dropdown as such?

sudar replied:
Let's just leave it like that.

sudar commented on include/Metabox/Posts/CustomTaxonomy.php:
```diff
@@ -0,0 +1,161 @@
+
+				<tr>
+					<td scope="row" colspan="2">
+						<select name="smbm_mbct_post_type" id="smbm_mbct_post_type">
```
Abstract this into a helper function that is added to the base class

mariadanieldeepak replied:
> Abstract this into a helper function that is added to the base class

Fixed

sudar commented on include/Metabox/Posts/CustomTaxonomy.php:
```diff
@@ -0,0 +1,161 @@
+		$wp_query    = new WP_Query();
+		$posts_count = 0;
+
+		if ( - 1 === $options['old_term'] ) {
```
Don't add a space between `-` and `1`. This will result in very bad readability.

mariadanieldeepak replied:
> Don't add a space between - and 1. This will result in very bad readability.

Fixed.

sudar commented on include/Request/CustomTaxonomyAction.php:
```diff
@@ -0,0 +1,84 @@
+		/* The action should be in the {page_slug}-nonce format.
+		 * {page_slug} is the slug of the page on which the AJAX is requested.
+		 */
+		check_ajax_referer( 'bulk-move-posts-nonce', 'nonce' );
```
Don't hardcode the nonce name. Build it dynamically from page slug. Also move this logic into a common function.

mariadanieldeepak replied:
> Don't hardcode the nonce name.

https://github.com/sudar/bulk-move/blob/43-feature/refactor-move-by-taxonomy/include/BulkMove.php#L179

From the above line, it looks like we have to retrieve the page slug by hard-coding the slug. This made me use the slug directly. Once we have a key to indicate different pages, we could refactor it.

> Build it dynamically from page slug. Also move this logic into a common function.

I removed the `check_admin_referrer()` in AJAX requests and used the one in `Page/Base.php`

https://github.com/sudar/bulk-move/blob/43-feature/refactor-move-by-taxonomy/include/Page/Base.php#L130-L154

mariadanieldeepak committed:
Abstract render post type helper in to Base class

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r158767544

mariadanieldeepak committed:
Use helper function to render Submit button

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r158767501

mariadanieldeepak committed:
Remove unwanted space character

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r158767584

mariadanieldeepak committed:
Add Bulk Move translation filter to be used from Metabox modules

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r158767396

mariadanieldeepak committed:
Remove unwanted blank line

mariadanieldeepak committed:
Remove unused methods

mariadanieldeepak committed:
Use BasePage nonce verifier for AJAX requests

mariadanieldeepak committed:
Fix StyleCI issues

mariadanieldeepak committed:
Fix StyleCI issues

mariadanieldeepak committed:
Merge branch 'dev/2.0.0' into 43-feature/refactor-move-by-taxonomy

# Conflicts:
#	include/BulkMove.php

mariadanieldeepak commented:
@sudar,

This is now ready for your review

sudar commented on include/Page/Base.php:
```diff
@@ -121,13 +121,14 @@ protected function register_hooks() {
 	 */
 	public function verify_nonce() {
-		if ( ! isset( $_POST['bm_action'] ) ) {
+		if ( ! isset( $_POST['bm_action'] ) || ! isset( $_POST['action'] ) ) {
```
Let's use `bm_action` itself.

mariadanieldeepak replied:
> Let's use bm_action itself.

Fixed by removing the other one.

sudar committed:
Merge branch 'dev/2.0.0' into 43-feature/refactor-move-by-taxonomy

sudar committed:
Load AJAX Handlers

sudar commented:
@mariadanieldeepak 

After further testing I found that using seperate classes for AJAX Handlers is inevitable.

I have now included them from Bulk Move.

Let's use this and remove the loadie implementation.

sudar commented on include/Page/Base.php:
```diff
@@ -232,19 +233,20 @@ public function enqueue_script() {
+		 *
+		 * @param array $translation_array A single dimensional translation array.
+		 */
+		$translation_array = apply_filters( 'bm_bulk-move_js_i18n', $translation_array );
```
Let's use the name `bm_js_translations` as the name of the filter.

mariadanieldeepak replied:
> Let's use the name bm_js_translations as the name of the filter.

Fixed and I sticked with the existing name that's used throughout the plugin.

mariadanieldeepak committed:
Merge branch 'dev/2.0.0' into 43-feature/refactor-move-by-taxonomy

mariadanieldeepak committed:
Remove breaking check from verify_nonce() method

mariadanieldeepak committed:
Remove Loadie implementation

mariadanieldeepak committed:
Remove old Bulk Move Posts class

mariadanieldeepak committed:
Fix Bulk Move by Custom Taxonomy ajax actions

mariadanieldeepak committed:
Remove Move posts by Custom Taxonomy button from JS validation

mariadanieldeepak committed:
Extend using the right class

mariadanieldeepak committed:
Fix Style CI issue

sudar commented on include/BulkMove.php:
```diff
@@ -255,6 +292,28 @@ public function get_translations_path() {
 		return $this->translations_path;
 	}
 
+	/**
```
Let's move Loadie implementation to a new branch.

mariadanieldeepak replied:
> Let's move Loadie implementation to a new branch.

Fixed.

sudar replied:
@mariadanieldeepak 

I am still seeing `add_loadie` method in the diff. Has it been removed?

mariadanieldeepak replied:
> I am still seeing add_loadie method in the diff. Has it been removed?

Ah, I missed to remove this method. The Interface was removed already. Fixed now.

sudar reviewed (changes requested):
Once we remove `Loadie` implementation this is good to go.

mariadanieldeepak committed:
Refactor JS

mariadanieldeepak committed:
Fix Remove Category not working in Bulk move by Category

mariadanieldeepak committed:
Fix Remove Tag by Bulk Move by Tag

mariadanieldeepak committed:
Merge branch 'dev/2.0.0' into 43-feature/refactor-move-by-taxonomy

mariadanieldeepak committed:
Fix Assert count

mariadanieldeepak committed:
Fix Assert count

mariadanieldeepak committed:
Fix Assert count

mariadanieldeepak commented:
@anandB000,

The following file uses a combination of tabs and spaces as Whitespaces.

https://github.com/sudar/bulk-move/blob/43-feature/refactor-move-by-taxonomy/tests/wp-tests/BM/Metabox/Posts/TagTest.php

For example, refer these line https://github.com/sudar/bulk-move/blob/43-feature/refactor-move-by-taxonomy/tests/wp-tests/BM/Metabox/Posts/TagTest.php#L111-L286

Let's get this fixed after @sudar merges #49. Otherwise these changes will create lot of noises in diff. I've created an [issue](https://github.com/sudar/bulk-move/issues/78) for you to work on this.

mariadanieldeepak committed:
Include Move by Custom Taxonomy tests

mariadanieldeepak committed:
Fix naming convention bug

mariadanieldeepak committed:
Use correct Base class to extend

mariadanieldeepak commented:
@sudar,

Kindly merge PR https://github.com/sudar/wp-plugin-test-tools/pull/2, before merging this one, as it is dependant on the other one.

Also, I've included #44 code changes as part of this PR, so we can close #44.

mariadanieldeepak committed:
Add Bulk Move by Users tests

mariadanieldeepak committed:
Validate Bulk Move by Custom Taxonomy on click

sudar commented on include/Metabox/Posts/Category.php:
```diff
@@ -90,14 +90,16 @@ public function move( $options ) {
 
 		foreach ( $posts as $post ) {
-			$current_cats = array_diff( wp_get_post_categories( $post->ID ), array( $options['old_cat'] ) );
+			$current_cats = wp_get_post_categories( $post->ID );
```
`wp_get_post_categories` can retrun `WP_Error`. We need to add an additional check.

_I think this was an existing issue and not related to your change_.

mariadanieldeepak replied:
> `wp_get_post_categories` can retrun `WP_Error`. We need to add an additional check.

Fixed.

sudar commented on include/Metabox/Posts/Tag.php:
```diff
@@ -68,23 +68,24 @@ public function move( $options ) {
 		foreach ( $posts as $post ) {
-            $current_tags = wp_get_post_tags( $post->ID, array( 'fields' => 'ids' ) );
-            $current_tags = array_diff( $current_tags, array( $options['old_tag'] ) );
+			$current_tags = wp_get_post_tags( $post->ID, array( 'fields' => 'ids' ) );
```
_Comment from Scrutinizer. Looks like an existing issue though_

It seems like $current_tags can also be of type WP_Error; however, parameter $array1 of array_diff() does only seem to accept array, maybe add an additional type check?

sudar commented on include/Request/CustomTaxonomyAction.php:
```diff
@@ -0,0 +1,102 @@
+class BM_Request_CustomTaxonomyAction {
+
+	/**
+	 * The BM_Loadie() calls this method.
```
Let's remove reference to BM_Loadie.

sudar commented:
@mariadanieldeepak 

The changes look good. I have just two more comments. One about the `add_loadie` method and the other a couple of catches by Scrutinizer.

Once they are fixed, we can merge it.

Also Travis seems to be failing. It think this may be because of the required changes to the base testing library. If yes, then include the commit hash in composer.json file to fix it.

mariadanieldeepak committed:
Remove add_loadie() method

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r167421613

mariadanieldeepak committed:
Handle WP_Error

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r167421758

mariadanieldeepak committed:
Remove irrelevant comment

Fix review comment: https://github.com/sudar/bulk-move/pull/49#discussion_r167421770

mariadanieldeepak committed:
Include composer.lock file

sudar commented:
@mariadanieldeepak 

Actually you have to modify only the `composer.json` not `composer.lock` file.

sudar committed:
Specify the commit hash for test tool

sudar committed:
Update the composer dependencies

sudar committed:
Fix the directory check for manually pulling test tools

sudar committed:
Use correct SPDX notation for license

sudar committed:
Downgrade to phpunit v6.0.0

phpunit 7.0.0 is not compatible with wp_mock yet.

sudar commented on include/Request/CustomTaxonomyAction.php:
```diff
@@ -0,0 +1,102 @@
+	 *
+	 * @since 2.0.0
+	 */
+	public static function load_custom_taxonomy_by_post_type() {
```
This should not be a `static` method

sudar commented on include/Request/CustomTaxonomyAction.php:
```diff
@@ -0,0 +1,102 @@
+	 *
+	 * @since 2.0.0
+	 */
+	public static function load_custom_terms_by_taxonomy() {
```
This should not be a `static` method

sudar committed:
Skip translation function

sudar committed:
Revert "Skip translation function"

This reverts commit 4f6b46e50fb23e86b599db07fe2efcabf0941b29.

sudar committed:
Fail travis build if one of the command in script fails

sudar committed:
Send code coverage reports only when the build is successful

sudar committed:
Revert "Revert "Skip translation function""

This reverts commit 99dc0eb31c2812af15bbe209d182a89df7a6c23b.

sudar committed:
Don't use static methods in AJAX Handler

sudar committed:
Run all travis commands as a single npm command

sudar committed:
Allow composer validate to have warnings

sudar commented:
@mariadanieldeepak 

I fixed the remaining issues and also composer/travis issues to get this PR merged.
